src/main_full.rs

```diff
@@ -66,96 +69,82 @@ pub fn main() {
     let targets =
         values_t!(matches.values_of("emit"), CompilationTarget).unwrap_or_else(|e| e.exit());
 
-    match compile(input_file, output_dir, targets) {
+    match compile_and_write(input_file, output_dir, targets) {
         Ok(_) => println!("Compiled {}. Outputs in {}", input_file, output_dir),
         Err(err) => println!("Unable to compile {}. \nError: {}", input_file, err),
     }
 }
 
-fn verify_nonexistent_or_empty(dir: &Path) -> Result<(), String> {
-    if !dir.exists() || dir.read_dir().map_err(ioerr_to_string)?.next().is_none() {
-        Ok(())
-    } else {
-        Err(format!("Directory '{}' is not empty", dir.display()))
-    }
-}
-
-fn compile(
+fn compile_and_write(
     src_file: &str,
     output_dir: &str,
     targets: Vec<CompilationTarget>,
 ) -> Result<(), String> {
-    let output_dir = Path::new(output_dir);
+    let src = fs::read_to_string(src_file).map_err(ioerr_to_string)?;
+    let compiled_module = fe_compiler::compile(&src).map_err(|error| error.to_string())?;
+
+    write_compiled_module(compiled_module, output_dir, targets.as_slice())
+}
 
+fn write_compiled_module(
+    mut module: CompiledModule,
+    output_dir: &str,
+    targets: &[CompilationTarget],
+) -> Result<(), String> {
+    let output_dir = Path::new(output_dir);
     verify_nonexistent_or_empty(output_dir)?;
     std::fs::create_dir_all(output_dir).map_err(ioerr_to_string)?;
 
-    let contents = fs::read_to_string(src_file).map_err(ioerr_to_string)?;
-
-    let output = fe_compiler::evm::compile(&contents, to_compile_stage(&targets))
```

Comment:
Just double checking since the compile stages seem to be gone. Can we still compile code that produces invalid YUL? E.g. with the current compiler one can compile `fe foo.fe --emit yul` even if `fe foo.fe` wouldn't work because the yul -> binary compilation errors. This is helpful to generate and inspect the faulty YUL code that fails to compile to binary.

Reply:
Good point :+1: